<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Oncoplot • ggalign</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Oncoplot">
<meta name="robots" content="noindex">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">ggalign</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">0.0.5.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/heatmap-layout.html">Heatmap Layout</a></li>
    <li><a class="dropdown-item" href="../articles/layout-customize.html">Layout Customization</a></li>
    <li><a class="dropdown-item" href="../articles/layout-plot.html">Layout Plot</a></li>
    <li><a class="dropdown-item" href="../articles/quad-layout.html">Quad-Layout</a></li>
    <li><a class="dropdown-item" href="../articles/stack-layout.html">Stack Layout</a></li>
    <li><a class="dropdown-item" href="../articles/operator.html">Operator</a></li>
    <li><a class="dropdown-item" href="../articles/plot-options.html">Plot options</a></li>
    <li><a class="dropdown-item" href="../articles/oncoplot.html">Oncoplot</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip> </h6></li>
    <li><a class="dropdown-item" href="../articles/difference-with-ggplot2.html">Difference with ggplot2</a></li>
    <li><a class="dropdown-item" href="../articles/complete-examples.html">Complete Examples</a></li>
    <li><a class="dropdown-item" href="../articles/plot-composer.html">Plot composer</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip> </h6></li>
    <li><a class="dropdown-item" href="../articles/FAQ.html">FAQ</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip> </h6></li>
    <li><a class="dropdown-item" href="../articles/performance.html">Performance</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>ComplexHeatmap</h6></li>
    <li><a class="dropdown-item" href="../articles/single-heatmap.html">ComplexHeatmap: single-heatmap</a></li>
    <li><a class="dropdown-item" href="../articles/more-examples.html">ComplexHeatmap: more examples</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Yunuuuu/ggalign/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Oncoplot</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Yunuuuu/ggalign/blob/main/vignettes/articles/oncoplot.Rmd" class="external-link"><code>vignettes/articles/oncoplot.Rmd</code></a></small>
      <div class="d-none name"><code>oncoplot.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="ggoncoplot">ggoncoplot<a class="anchor" aria-label="anchor" href="#ggoncoplot"></a>
</h2>
<p>The <code><a href="../reference/ggoncoplot.html">ggoncoplot()</a></code> function generates
<code>oncoPrint</code> visualizations that display genetic alterations
in a matrix format. This function is especially useful for visualizing
complex genomic data, such as mutations, copy number variations, and
other genomic alterations in cancer research.</p>
</div>
<div class="section level2">
<h2 id="input-data">Input data<a class="anchor" aria-label="anchor" href="#input-data"></a>
</h2>
<p>The input should be a character matrix which encodes the alterations,
you can use string of <code>";"</code>, <code>":"</code>,
<code>","</code>, and <code>"|"</code> to separate multiple alterations.
Internally, <code><a href="../reference/ggoncoplot.html">ggoncoplot()</a></code> will use
<code><a href="../reference/fortify_matrix.html">fortify_matrix()</a></code> to get such matrix.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/textconnections.html" class="external-link">textConnection</a></span><span class="op">(</span></span>
<span>        <span class="st">"s1,s2,s3</span></span>
<span><span class="st">         g1,snv;indel,snv,indel</span></span>
<span><span class="st">         g2,,snv;indel,snv</span></span>
<span><span class="st">         g3,snv,,indel;snv"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    row.names <span class="op">=</span> <span class="fl">1</span>, header <span class="op">=</span> <span class="cn">TRUE</span>, sep <span class="op">=</span> <span class="st">","</span>, stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">mat</span></span>
<span><span class="co">#&gt;                    s1        s2        s3</span></span>
<span><span class="co">#&gt;          g1 snv;indel       snv     indel</span></span>
<span><span class="co">#&gt;          g2           snv;indel       snv</span></span>
<span><span class="co">#&gt;          g3       snv           indel;snv</span></span></code></pre></div>
<p>A basic oncoprint can be generated as follows:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Yunuuuu/ggalign" class="external-link">ggalign</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: ggplot2</span></span>
<span><span class="fu"><a href="../reference/ggoncoplot.html">ggoncoplot</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span></span></code></pre></div>
<p><img src="oncoplot_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="oncoprint-customization">oncoPrint Customization<a class="anchor" aria-label="anchor" href="#oncoprint-customization"></a>
</h2>
<p>By default, all alterations are represented with tiles of equal width
and height, which may lead to overlapping. You can control the
<code>width</code> and <code>height</code> of the tiles using the
<code>map_width</code> and <code>map_height</code> arguments (we will
introduce another more effective ways to handle this in the
<code>Specialized Geoms</code> section):</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ggoncoplot.html">ggoncoplot</a></span><span class="op">(</span><span class="va">mat</span>, map_width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>snv <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span>, map_height <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>indel <span class="op">=</span> <span class="fl">0.9</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="oncoplot_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p>By default, all empty string will be converted to <code>NA</code>
value and <code>ggplot2</code> will translate the <code>NA</code> values
and render it in the legend. To prevent this, you can turn off the
translation using <code>na.translate = FALSE</code>:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ggoncoplot.html">ggoncoplot</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dark2"</span>, na.translate <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="oncoplot_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p>The function automatically reorders rows and columns based on
alteration frequency. Disable this with <code>reorder_row = FALSE</code>
and <code>reorder_column = FALSE</code>.</p>
<p>You can further customize layouts using methods from <a href="https://yunuuuu.github.io/ggalign/dev/articles/layout-customize.html">layout-customize</a>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ggoncoplot.html">ggoncoplot</a></span><span class="op">(</span><span class="va">mat</span>, reorder_row <span class="op">=</span> <span class="cn">FALSE</span>, reorder_column <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dark2"</span>, na.translate <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="oncoplot_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="advanced-data-handling">Advanced Data Handling<a class="anchor" aria-label="anchor" href="#advanced-data-handling"></a>
</h2>
<p><code><a href="../reference/ggoncoplot.html">ggoncoplot()</a></code> is built on <code><a href="../reference/heatmap_layout.html">ggheatmap()</a></code> with a
default <a href="https://yunuuuu.github.io/ggalign/dev/articles/plot-options.html"><code>plot_data()</code></a>,
which splits alterations into separate entries for visualization. (See
<code>ggplot2 specification</code> section in <code><a href="../reference/heatmap_layout.html">ggheatmap()</a></code>
for the input data of <code><a href="../reference/plot_data.html">plot_data()</a></code>):</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pdata</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/separate_longer_delim.html" class="external-link">separate_longer_delim</a></span><span class="op">(</span><span class="va">data</span>,</span>
<span>        cols <span class="op">=</span> <span class="va">value</span>,</span>
<span>        delim <span class="op">=</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/modifiers.html" class="external-link">regex</a></span><span class="op">(</span><span class="st">"\\s*[;:,|]\\s*"</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co"># Note: `ggoncoplot()` will automatically convert any empty strings to `NA`</span></span>
<span><span class="co">#        before pass it to `ggheatmap`. So this figure will contain an empty</span></span>
<span><span class="co">#        string value.</span></span>
<span><span class="fu"><a href="../reference/heatmap_layout.html">ggheatmap</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span> <span class="op">-</span></span>
<span>    <span class="fu"><a href="../reference/plot_data.html">plot_data</a></span><span class="op">(</span><span class="va">pdata</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dark2"</span>, na.translate <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; → heatmap built with `geom_tile()`</span></span></code></pre></div>
<p><img src="oncoplot_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>By default, the <code><a href="../reference/plot_data.html">plot_data()</a></code> is inherited from the parent
layout if the data of the plot is inherited from the layout. You can
apply the parent <code><a href="../reference/plot_data.html">plot_data()</a></code> first and then apply another
transformation by setting <code>inherit = TRUE</code>. This
functionality is especially useful when working with
<code><a href="../reference/ggoncoplot.html">ggoncoplot()</a></code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ggoncoplot.html">ggoncoplot</a></span><span class="op">(</span><span class="va">mat</span>, map_width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>snv <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span>, map_height <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>indel <span class="op">=</span> <span class="fl">0.9</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="../reference/quad_active.html">anno_top</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="../reference/align_gg.html">ggalign</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="co"># by setting `inherti = TRUE`, we apply the parent layout `plot_data()`</span></span>
<span>    <span class="co"># (from the `ggoncoplot()` layout) firstly, which will split the alteration</span></span>
<span>    <span class="co"># string and unnested the columns.</span></span>
<span>    <span class="co"># Here: We then remove `NA` value</span></span>
<span>    <span class="fu"><a href="../reference/plot_data.html">plot_data</a></span><span class="op">(</span><span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">.x</span>, <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">value</span><span class="op">)</span><span class="op">)</span>, inherit <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">.x</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes_eval.html" class="external-link">after_stat</a></span><span class="op">(</span><span class="va">count</span><span class="op">)</span>, fill <span class="op">=</span> <span class="va">value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="co"># note: `ggoncoplot()` use `geom_tile()` to draw the oncoPrint,</span></span>
<span>    <span class="co"># the guide is different from `geom_bar()`, though both looks</span></span>
<span>    <span class="co"># like the same, we remove the guide of `geom_bar()` plot</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dark2"</span>, na.translate <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="oncoplot_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="integration-with-maftools">Integration with maftools<a class="anchor" aria-label="anchor" href="#integration-with-maftools"></a>
</h2>
<p>The <code><a href="../reference/ggoncoplot.html">ggoncoplot()</a></code> function supports <code>MAF</code>
objects from the <code>maftools</code> package using the
<code><a href="../reference/fortify_matrix.MAF.html">fortify_matrix.MAF()</a></code> method. It includes options to control
data generation for <code><a href="../reference/ggoncoplot.html">ggoncoplot()</a></code>, such as drawing the top
<code>n_top</code> genes.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load data from `maftools`</span></span>
<span><span class="va">laml.maf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"tcga_laml.maf.gz"</span>, package <span class="op">=</span> <span class="st">"maftools"</span><span class="op">)</span></span>
<span><span class="co"># clinical information containing survival information and histology. This is optional</span></span>
<span><span class="va">laml.clin</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"tcga_laml_annot.tsv"</span>, package <span class="op">=</span> <span class="st">"maftools"</span><span class="op">)</span></span>
<span><span class="va">laml</span> <span class="op">&lt;-</span> <span class="fu">maftools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/maftools/man/read.maf.html" class="external-link">read.maf</a></span><span class="op">(</span></span>
<span>    maf <span class="op">=</span> <span class="va">laml.maf</span>,</span>
<span>    clinicalData <span class="op">=</span> <span class="va">laml.clin</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Visualizing the Top 20 Genes</span></span>
<span><span class="fu"><a href="../reference/ggoncoplot.html">ggoncoplot</a></span><span class="op">(</span><span class="va">laml</span>, n_top <span class="op">=</span> <span class="fl">20</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dark2"</span>, na.translate <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="../reference/theme_no_axes.html">theme_no_axes</a></span><span class="op">(</span><span class="st">"x"</span><span class="op">)</span></span></code></pre></div>
<p><img src="oncoplot_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<p>By default, samples without alterations in the selected genes are
removed. To include all samples, set
<code>remove_empty_samples = FALSE</code>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/align_plots.html">align_plots</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="../reference/ggoncoplot.html">ggoncoplot</a></span><span class="op">(</span><span class="va">laml</span>, n_top <span class="op">=</span> <span class="fl">20L</span>, remove_empty_samples <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dark2"</span>, na.translate <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="../reference/theme_no_axes.html">theme_no_axes</a></span><span class="op">(</span><span class="st">"x"</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Keep empty samples"</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="../reference/ggoncoplot.html">ggoncoplot</a></span><span class="op">(</span><span class="va">laml</span>, n_top <span class="op">=</span> <span class="fl">20L</span>, remove_empty_samples <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dark2"</span>, na.translate <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="../reference/theme_no_axes.html">theme_no_axes</a></span><span class="op">(</span><span class="st">"x"</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Remove empty samples"</span><span class="op">)</span>,</span>
<span>    ncol <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>    guides <span class="op">=</span> <span class="st">"tlbr"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="oncoplot_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="specialized-geoms">Specialized Geoms<a class="anchor" aria-label="anchor" href="#specialized-geoms"></a>
</h2>
<div class="section level3">
<h3 id="geom_subtile">geom_subtile()<a class="anchor" aria-label="anchor" href="#geom_subtile"></a>
</h3>
<p>When multiple alterations occur in the same sample and gene, they are
combined into a single value, <code>"Multi_Hit"</code>, by default. To
visualize these alterations separately, you can set
<code>collapse_vars = FALSE</code>. However, doing so can lead to
overlapping alterations within the same cell, making the visualization
cluttered and hard to interpret.</p>
<p>In such cases, disabling the default filling and defining a custom
heatmap layer with <code><a href="../reference/geom_subrect.html">geom_subtile()</a></code> is more effective. This
function subdivides each cell into smaller rectangles, allowing the
distinct alterations to be clearly displayed.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ggoncoplot.html">ggoncoplot</a></span><span class="op">(</span><span class="va">laml</span>, n_top <span class="op">=</span> <span class="fl">20</span>, collapse_vars <span class="op">=</span> <span class="cn">FALSE</span>, filling <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="../reference/geom_subrect.html">geom_subtile</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">value</span><span class="op">)</span>, direction <span class="op">=</span> <span class="st">"vertical"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dark2"</span>, na.translate <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="../reference/theme_no_axes.html">theme_no_axes</a></span><span class="op">(</span><span class="st">"x"</span><span class="op">)</span></span>
<span><span class="co">#&gt; `geom_subtile()` subdivide tile into a maximal of 3 rectangles</span></span></code></pre></div>
<p><img src="oncoplot_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<p>To focus only on cells with multiple alterations, filter the data
before applying <code><a href="../reference/geom_subrect.html">geom_subtile()</a></code>:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ggoncoplot.html">ggoncoplot</a></span><span class="op">(</span><span class="va">laml</span>, n_top <span class="op">=</span> <span class="fl">20</span>, collapse_vars <span class="op">=</span> <span class="cn">FALSE</span>, filling <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="../reference/plot_data.html">plot_data</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">data</span>, <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">1L</span>, .by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">".x"</span>, <span class="st">".y"</span><span class="op">)</span><span class="op">)</span></span>
<span>        <span class="co"># we apply the parent layout `plot_data()` first (`inherit = TRUE`),</span></span>
<span>        <span class="co"># which will split the alteration string and unnested the columns.</span></span>
<span>    <span class="op">}</span>, inherit <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="../reference/geom_subrect.html">geom_subtile</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">value</span><span class="op">)</span>, direction <span class="op">=</span> <span class="st">"vertical"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dark2"</span>, na.translate <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="../reference/theme_no_axes.html">theme_no_axes</a></span><span class="op">(</span><span class="st">"x"</span><span class="op">)</span></span>
<span><span class="co">#&gt; `geom_subtile()` subdivide tile into a maximal of 3 rectangles</span></span></code></pre></div>
<p><img src="oncoplot_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="geom_draw">geom_draw()<a class="anchor" aria-label="anchor" href="#geom_draw"></a>
</h3>
<p><code><a href="../reference/geom_subrect.html">geom_subtile()</a></code> often suffices for most scenarios.
However, if you require a strategy similar to that of
<code>ComplexHeatmap</code>, consider using <code><a href="../reference/geom_draw.html">geom_draw()</a></code>,
which offers greater flexibility for complex customizations.</p>
<p>Key Features of <code><a href="../reference/geom_draw.html">geom_draw()</a></code>:</p>
<ul>
<li>Custom Drawing Logic: Accepts a <code>draw</code> aesthetic,
allowing each cell to be rendered as a specific graphical object
(<code>grob</code>).</li>
<li>Manual Scaling: Requires <code><a href="../reference/scale_draw_manual.html">scale_draw_manual()</a></code> to map
<code>draw</code> values to corresponding drawing functions.</li>
<li>Flexible Aesthetic Mapping: Functions mapped through
<code><a href="../reference/scale_draw_manual.html">scale_draw_manual()</a></code> can utilize any number of ggplot2
aesthetics and output custom graphical objects (<code>grob</code>).
Beyond the <code>x</code>, <code>y</code>, <code>xmin</code>,
<code>xmax</code>, <code>width</code>, and <code>height</code>
aesthetics, you may want to rely solely on mapped aesthetics
(<code><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes()</a></code>). Unmapped aesthetics will default to ggplot2’s
behavior, which might not produce the desired outcome. And always use
<code>native</code> unit.</li>
</ul>
<p>:) Sorry, I cannot deal with a pretty example for this, but the
function provided in <code>values</code> argument of
<code>scale_draw_manual</code> operates similarly to
<code>alter_fun</code> in ComplexHeatmap. Currently,
<code><a href="../reference/geom_subrect.html">geom_subtile()</a></code> performs well enough for most cases, making
<code><a href="../reference/geom_draw.html">geom_draw()</a></code> somewhat cumbersome to use. Managing alteration
types, especially in the case of overlapping alterations, requires
meticulous design to ensure that the visual elements do not interfere
with each other.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">grid</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>value <span class="op">=</span> <span class="va">letters</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span><span class="op">]</span>, y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="../reference/geom_draw.html">geom_draw</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fl">1</span>, y <span class="op">=</span> <span class="va">y</span>, draw <span class="op">=</span> <span class="va">value</span>, fill <span class="op">=</span> <span class="va">value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="../reference/scale_draw_manual.html">scale_draw_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        a <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">width</span>, <span class="va">height</span>, <span class="va">fill</span><span class="op">)</span> <span class="op">{</span></span>
<span>            <span class="fu"><a href="https://rdrr.io/r/grid/grid.rect.html" class="external-link">rectGrob</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>,</span>
<span>                width <span class="op">=</span> <span class="va">width</span>, height <span class="op">=</span> <span class="va">height</span>,</span>
<span>                gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">fill</span><span class="op">)</span>,</span>
<span>                default.units <span class="op">=</span> <span class="st">"native"</span></span>
<span>            <span class="op">)</span></span>
<span>        <span class="op">}</span>,</span>
<span>        b <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">width</span>, <span class="va">height</span>, <span class="va">fill</span><span class="op">)</span> <span class="op">{</span></span>
<span>            <span class="fu"><a href="https://rdrr.io/r/grid/grid.rect.html" class="external-link">rectGrob</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>,</span>
<span>                width <span class="op">=</span> <span class="va">width</span>, height <span class="op">=</span> <span class="va">height</span>,</span>
<span>                gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">fill</span><span class="op">)</span>,</span>
<span>                default.units <span class="op">=</span> <span class="st">"native"</span></span>
<span>            <span class="op">)</span></span>
<span>        <span class="op">}</span>,</span>
<span>        c <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">width</span>, <span class="va">height</span>, <span class="va">fill</span><span class="op">)</span> <span class="op">{</span></span>
<span>            <span class="fu"><a href="https://rdrr.io/r/grid/grid.rect.html" class="external-link">rectGrob</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>,</span>
<span>                width <span class="op">=</span> <span class="va">width</span>, height <span class="op">=</span> <span class="va">height</span>,</span>
<span>                gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">fill</span><span class="op">)</span>,</span>
<span>                default.units <span class="op">=</span> <span class="st">"native"</span></span>
<span>            <span class="op">)</span></span>
<span>        <span class="op">}</span>,</span>
<span>        d <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">width</span>, <span class="va">height</span>, <span class="va">shape</span><span class="op">)</span> <span class="op">{</span></span>
<span>            <span class="fu"><a href="https://rdrr.io/r/grid/grid.grob.html" class="external-link">gList</a></span><span class="op">(</span></span>
<span>                <span class="fu"><a href="https://rdrr.io/r/grid/grid.points.html" class="external-link">pointsGrob</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, pch <span class="op">=</span> <span class="va">shape</span><span class="op">)</span>,</span>
<span>                <span class="co"># To ensure the rectangle color is shown in the legends, you</span></span>
<span>                <span class="co"># must explicitly provide a color argument and include it in the</span></span>
<span>                <span class="co"># `gpar()` of the graphical object</span></span>
<span>                <span class="fu"><a href="https://rdrr.io/r/grid/grid.rect.html" class="external-link">rectGrob</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">width</span>, <span class="va">height</span>,</span>
<span>                    gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>col <span class="op">=</span> <span class="st">"black"</span>, fill <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span>,</span>
<span>                    default.units <span class="op">=</span> <span class="st">"native"</span></span>
<span>                <span class="op">)</span></span>
<span>            <span class="op">)</span></span>
<span>        <span class="op">}</span>,</span>
<span>        e <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">xmin</span>, <span class="va">xmax</span>, <span class="va">ymin</span>, <span class="va">ymax</span><span class="op">)</span> <span class="op">{</span></span>
<span>            <span class="fu"><a href="https://rdrr.io/r/grid/grid.segments.html" class="external-link">segmentsGrob</a></span><span class="op">(</span></span>
<span>                <span class="va">xmin</span>,</span>
<span>                <span class="va">ymin</span>,</span>
<span>                <span class="va">xmax</span>,</span>
<span>                <span class="va">ymax</span>,</span>
<span>                gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">gpar</a></span><span class="op">(</span>lwd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>,</span>
<span>                default.units <span class="op">=</span> <span class="st">"native"</span></span>
<span>            <span class="op">)</span></span>
<span>        <span class="op">}</span></span>
<span>    <span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dark2"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="oncoplot_files/figure-html/unnamed-chunk-15-1.png" width="192"></p>
</div>
</div>
<div class="section level2">
<h2 id="ggalign-attributes">ggalign attributes<a class="anchor" aria-label="anchor" href="#ggalign-attributes"></a>
</h2>
<p>Except for the data used for the main plot,
<code><a href="../reference/fortify_matrix.MAF.html">fortify_matrix.MAF()</a></code> also attaches several useful
attributes.</p>
<ul>
<li>
<code>gene_summary</code>: gene summary informations. See
<code><a href="https://rdrr.io/pkg/maftools/man/getGeneSummary.html" class="external-link">maftools::getGeneSummary()</a></code> for details.</li>
<li>
<code>sample_summary</code>: sample summary informations. See
<code><a href="https://rdrr.io/pkg/maftools/man/getSampleSummary.html" class="external-link">maftools::getSampleSummary()</a></code> for details.</li>
<li>
<code>sample_anno</code>: sample clinical informations. See
<code><a href="https://rdrr.io/pkg/maftools/man/getClinicalData.html" class="external-link">maftools::getClinicalData()</a></code> for details.</li>
<li>
<code>n_genes</code>: Total of genes.</li>
<li>
<code>n_samples</code>: Total of samples.</li>
<li>
<code>titv</code>: A list of <code>data.frames</code> with
Transitions and Transversions summary. See <code><a href="https://rdrr.io/pkg/maftools/man/titv.html" class="external-link">maftools::titv()</a></code>
for details.</li>
</ul>
<p>You can extract these attributes using the
<code><a href="../reference/ggalign_attr.html">ggalign_attr()</a></code> function. Below is an example of how to
extract the sample summary data (e.g., TMB) using
<code><a href="../reference/ggalign_attr.html">ggalign_attr()</a></code> and add it to the top annotation of an
oncoplot:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ggoncoplot.html">ggoncoplot</a></span><span class="op">(</span><span class="va">laml</span>, n_top <span class="op">=</span> <span class="fl">20</span>, collapse_vars <span class="op">=</span> <span class="cn">FALSE</span>, filling <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="../reference/geom_subrect.html">geom_subtile</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">value</span><span class="op">)</span>, direction <span class="op">=</span> <span class="st">"vertical"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="../reference/theme_no_axes.html">theme_no_axes</a></span><span class="op">(</span><span class="st">"x"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="../reference/quad_active.html">anno_top</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="../reference/align_gg.html">ggalign</a></span><span class="op">(</span>data <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ggalign_attr.html">ggalign_attr</a></span><span class="op">(</span><span class="va">data</span>, <span class="st">"sample_summary"</span><span class="op">)</span></span>
<span>        <span class="co"># matrix input will be automatically melted into a long foramted data</span></span>
<span>        <span class="co"># frame in `ggalign()` function.</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">data</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1L</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="op">}</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">.x</span>, <span class="va">value</span>, fill <span class="op">=</span> <span class="va">.column_names</span><span class="op">)</span>,</span>
<span>        stat <span class="op">=</span> <span class="st">"identity"</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"TMB"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dark2"</span>, na.translate <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; `geom_subtile()` subdivide tile into a maximal of 3 rectangles</span></span></code></pre></div>
<p><img src="oncoplot_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
<p>We can draw the example from <a href="https://bioconductor.org/packages/release/bioc/vignettes/maftools/inst/doc/oncoplots.html" class="external-link">maftools
vignette</a>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ggoncoplot.html">ggoncoplot</a></span><span class="op">(</span><span class="va">laml</span>, n_top <span class="op">=</span> <span class="fl">20</span>, collapse_vars <span class="op">=</span> <span class="cn">FALSE</span>, filling <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="../reference/geom_subrect.html">geom_subtile</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">value</span><span class="op">)</span>, direction <span class="op">=</span> <span class="st">"vertical"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="../reference/theme_no_axes.html">theme_no_axes</a></span><span class="op">(</span><span class="st">"x"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="co"># since legends from geom_tile (oncoPrint body) and `geom_bar`</span></span>
<span>    <span class="co"># is different, though both looks like the same, the internal</span></span>
<span>    <span class="co"># won't merge the legends. we remove the legends of oncoPrint body</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="co"># add top annotation</span></span>
<span>    <span class="fu"><a href="../reference/quad_active.html">anno_top</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="../reference/align_gg.html">ggalign</a></span><span class="op">(</span>data <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ggalign_attr.html">ggalign_attr</a></span><span class="op">(</span><span class="va">data</span>, <span class="st">"sample_summary"</span><span class="op">)</span></span>
<span>        <span class="co"># matrix input will be automatically melted into a long foramted data</span></span>
<span>        <span class="co"># frame in `ggalign()` function.</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">data</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1L</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="op">}</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">.x</span>, <span class="va">value</span>, fill <span class="op">=</span> <span class="va">.column_names</span><span class="op">)</span>,</span>
<span>        stat <span class="op">=</span> <span class="st">"identity"</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"TMB"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="co"># add right annotation</span></span>
<span>    <span class="fu"><a href="../reference/quad_active.html">anno_right</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">-</span></span>
<span>    <span class="co"># remove bottom spaces of the right annotation when aligning</span></span>
<span>    <span class="fu"><a href="../reference/plot_align.html">plot_align</a></span><span class="op">(</span>free_spaces <span class="op">=</span> <span class="st">"b"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="co"># add the text percent for the alterated samples in the right annotation</span></span>
<span>    <span class="fu"><a href="../reference/align_gg.html">ggalign</a></span><span class="op">(</span>data <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="co"># Atomic vector will be put in the `value` column of the data frame.</span></span>
<span>        <span class="fu"><a href="../reference/ggalign_attr.html">ggalign_attr</a></span><span class="op">(</span><span class="va">data</span>, <span class="st">"gene_summary"</span><span class="op">)</span><span class="op">$</span><span class="va">AlteredSamples</span> <span class="op">/</span></span>
<span>            <span class="fu"><a href="../reference/ggalign_attr.html">ggalign_attr</a></span><span class="op">(</span><span class="va">data</span>, <span class="st">"n_samples"</span><span class="op">)</span></span>
<span>    <span class="op">}</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_text</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="fl">1</span>, label <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/label_percent.html" class="external-link">label_percent</a></span><span class="op">(</span><span class="op">)</span><span class="op">(</span><span class="va">value</span><span class="op">)</span><span class="op">)</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_x_continuous</a></span><span class="op">(</span></span>
<span>        expand <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expansion.html" class="external-link">expansion</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        name <span class="op">=</span> <span class="cn">NULL</span>, breaks <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>        limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">margin</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="co"># add the bar plot in the right annotation</span></span>
<span>    <span class="fu"><a href="../reference/align_gg.html">ggalign</a></span><span class="op">(</span>data <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ggalign_attr.html">ggalign_attr</a></span><span class="op">(</span><span class="va">data</span>, <span class="st">"gene_summary"</span><span class="op">)</span></span>
<span>        <span class="co"># matrix input will be automatically melted into a long foramted data</span></span>
<span>        <span class="co"># frame in `ggalign()` function.</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">data</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fl">8</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="op">}</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">value</span>, fill <span class="op">=</span> <span class="va">.column_names</span><span class="op">)</span>,</span>
<span>        stat <span class="op">=</span> <span class="st">"identity"</span>,</span>
<span>        orientation <span class="op">=</span> <span class="st">"y"</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"No. of samples"</span><span class="op">)</span> <span class="op">-</span></span>
<span>    <span class="co"># we apply the scale mapping to the top and right annotation: `position = "tr"`</span></span>
<span>    <span class="co"># and the main plot: `main = TRUE`</span></span>
<span>    <span class="fu"><a href="../reference/with_quad.html">with_quad</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_brewer</a></span><span class="op">(</span><span class="st">"Mutations"</span>,</span>
<span>            palette <span class="op">=</span> <span class="st">"Dark2"</span>, na.translate <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>        <span class="op">)</span>,</span>
<span>        position <span class="op">=</span> <span class="st">"tr"</span>,</span>
<span>        main <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="co"># add bottom annotation</span></span>
<span>    <span class="fu"><a href="../reference/quad_active.html">anno_bottom</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="co"># add bar plot in the bottom annotation</span></span>
<span>    <span class="fu"><a href="../reference/align_gg.html">ggalign</a></span><span class="op">(</span>data <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ggalign_attr.html">ggalign_attr</a></span><span class="op">(</span><span class="va">data</span>, <span class="st">"titv"</span><span class="op">)</span><span class="op">$</span><span class="va">fraction.contribution</span></span>
<span>        <span class="co"># matrix input will be automatically melted into a long foramted data</span></span>
<span>        <span class="co"># frame in `ggalign()` function.</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">data</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fl">7</span><span class="op">]</span><span class="op">)</span></span>
<span>    <span class="op">}</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">value</span>, fill <span class="op">=</span> <span class="va">.column_names</span><span class="op">)</span>, stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Ti/Tv"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_fill_brewer</a></span><span class="op">(</span><span class="st">"Ti/Tv"</span>, palette <span class="op">=</span> <span class="st">"Set2"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 24 rows containing missing values or values outside the scale range</span></span>
<span><span class="co">#&gt; (`geom_bar()`).</span></span>
<span><span class="co">#&gt; `geom_subtile()` subdivide tile into a maximal of 3 rectangles</span></span></code></pre></div>
<p><img src="oncoplot_files/figure-html/unnamed-chunk-17-1.png" width="1152"></p>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.2 (2024-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.5 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] grid      stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] ggalign_0.0.5.9000 ggplot2_3.5.1     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] sass_0.4.9         utf8_1.2.4         generics_0.1.3     tidyr_1.3.1       </span></span>
<span><span class="co">#&gt;  [5] maftools_2.22.0    lattice_0.22-6     stringi_1.8.4      digest_0.6.37     </span></span>
<span><span class="co">#&gt;  [9] magrittr_2.0.3     evaluate_1.0.1     RColorBrewer_1.1-3 fastmap_1.2.0     </span></span>
<span><span class="co">#&gt; [13] R.oo_1.27.0        Matrix_1.7-1       jsonlite_1.8.9     R.utils_2.12.3    </span></span>
<span><span class="co">#&gt; [17] survival_3.7-0     purrr_1.0.2        fansi_1.0.6        scales_1.3.0      </span></span>
<span><span class="co">#&gt; [21] textshaping_0.4.0  jquerylib_0.1.4    cli_3.6.3          rlang_1.1.4       </span></span>
<span><span class="co">#&gt; [25] R.methodsS3_1.8.2  munsell_0.5.1      splines_4.4.2      withr_3.0.2       </span></span>
<span><span class="co">#&gt; [29] cachem_1.1.0       yaml_2.3.10        DNAcopy_1.80.0     tools_4.4.2       </span></span>
<span><span class="co">#&gt; [33] dplyr_1.1.4        colorspace_2.1-1   vctrs_0.6.5        R6_2.5.1          </span></span>
<span><span class="co">#&gt; [37] lifecycle_1.0.4    stringr_1.5.1      fs_1.6.5           ragg_1.3.3        </span></span>
<span><span class="co">#&gt; [41] pkgconfig_2.0.3    desc_1.4.3         pkgdown_2.1.1      pillar_1.9.0      </span></span>
<span><span class="co">#&gt; [45] bslib_0.8.0        gtable_0.3.6       glue_1.8.0         data.table_1.16.2 </span></span>
<span><span class="co">#&gt; [49] systemfonts_1.1.0  xfun_0.49          tibble_3.2.1       tidyselect_1.2.1  </span></span>
<span><span class="co">#&gt; [53] knitr_1.49         farver_2.1.2       htmltools_0.5.8.1  rmarkdown_2.29    </span></span>
<span><span class="co">#&gt; [57] labeling_0.4.3     compiler_4.4.2</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Yun Peng.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
