<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-GB"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Create a Circular Layout for Genomic Data — circle_genomic • ggalign</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Create a Circular Layout for Genomic Data — circle_genomic"><meta name="description" content="circle_genomic() constructs a circular layout specifically for genomic
data. It is a specialized variant of circle_continuous() that applies
default axis limits and coerces the first column of each plot's data to use
chromosome (seqname) identifiers-matching those in the layout data-as
factor levels."><meta property="og:description" content="circle_genomic() constructs a circular layout specifically for genomic
data. It is a specialized variant of circle_continuous() that applies
default axis limits and coerces the first column of each plot's data to use
chromosome (seqname) identifiers-matching those in the layout data-as
factor levels."><meta property="og:image" content="https://yunuuuu.github.io/ggalign/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">ggalign</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.2.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="https://yunuuuu.github.io/ggalign-book/" aria-label="Tutorial"><span class="fa fa-solid fa-book"></span> Tutorial</a></li>
<li class="nav-item"><a class="nav-link" href="https://yunuuuu.github.io/ggalign-gallery/" aria-label="Gallery"><span class="fa fa-solid fa-grip"></span> Gallery</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/benchmarks.html"><span class="fa fa-solid fa-chart-line"></span> Benchmarks</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/FAQ.html"><span class="fa fa-solid fa-clipboard-question"></span> FAQ</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Yunuuuu/ggalign/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Create a Circular Layout for Genomic Data</h1>
      <small class="dont-index">Source: <a href="https://github.com/Yunuuuu/ggalign/blob/main/R/layout-chain-circle-genomic.R" class="external-link"><code>R/layout-chain-circle-genomic.R</code></a></small>
      <div class="d-none name"><code>circle_genomic.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>circle_genomic()</code> constructs a circular layout specifically for genomic
data. It is a specialized variant of <code><a href="circle_layout.html">circle_continuous()</a></code> that applies
default axis limits and coerces the first column of each plot's data to use
chromosome (<code>seqname</code>) identifiers-matching those in the layout data-as
factor levels.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">circle_genomic</span><span class="op">(</span></span>
<span>  <span class="va">data</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  radial <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  direction <span class="op">=</span> <span class="st">"outward"</span>,</span>
<span>  sector_spacing <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  theme <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-data">data<a class="anchor" aria-label="anchor" href="#arg-data"></a></dt>
<dd><p>The input data, which can be:</p><ul><li><p>A <code>character</code> string ("hg19" or "hg38") to load a predefined cytoband
reference.</p></li>
<li><p>A <code>data.frame</code> with at least three columns: <code>chromosome</code>, <code>start</code>, and
<code>end</code> positions.</p></li>
<li><p>A genomic object convertible via <code><a href="fortify_data_frame.html">fortify_data_frame()</a></code>.</p></li>
</ul></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Additional arguments passed to specific methods or
<code><a href="fortify_data_frame.html">fortify_data_frame()</a></code>.</p></dd>


<dt id="arg-radial">radial<a class="anchor" aria-label="anchor" href="#arg-radial"></a></dt>
<dd><p>A <code><a href="coord_circle.html">coord_circle()</a></code>/<code><a href="https://ggplot2.tidyverse.org/reference/coord_radial.html" class="external-link">coord_radial()</a></code>
object that defines the global parameters for coordinate across all plots
in the layout. The parameters <code>start</code>, <code>end</code>, <code>direction</code>, and <code>expand</code> will
be inherited and applied uniformly to all plots within the layout. The
parameters <code>theta</code> and <code>r.axis.inside</code> will always be ignored and will be set
to <code>"x"</code> and <code>TRUE</code>, respectively, for all plots.</p></dd>


<dt id="arg-direction">direction<a class="anchor" aria-label="anchor" href="#arg-direction"></a></dt>
<dd><p>A single string of <code>"inward"</code> or <code>"outward"</code>,
indicating the direction in which the plot is added.</p><ul><li><p><code>outward</code>: The plot is added from the inner to the outer.</p></li>
<li><p><code>inward</code>: The plot is added from the outer to the inner.</p></li>
</ul></dd>


<dt id="arg-sector-spacing">sector_spacing<a class="anchor" aria-label="anchor" href="#arg-sector-spacing"></a></dt>
<dd><p>The size of spacing between different panel. A numeric
of the radians or a <code><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">rel()</a></code> object.</p></dd>


<dt id="arg-theme">theme<a class="anchor" aria-label="anchor" href="#arg-theme"></a></dt>
<dd><p>A <code><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme()</a></code> object used to customize various
elements of the layout, including <code>guides</code>, <code>title</code>, <code>subtitle</code>, <code>caption</code>,
<code>margins</code>, <code>panel.border</code>, and <code>background</code>. By default, the theme will
inherit from the parent <code>layout</code>. It also controls the panel spacing for all
plots in the layout.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A <code>circle_layout</code> object representing the genomic layout.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Yun Peng, Shixiang Wang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

